lockup-factory/src/lib.rs

```diff
@@ -0,0 +1,423 @@
+    /// The base for the execution and the base for cash rollback.
+    pub const CALLBACK: Gas = BASE;
+}
+
+const MIN_ATTACHED_BALANCE: Balance = 35_000_000_000_000_000_000_000_000;
+
+/// External interface for the callbacks to self.
+#[ext_contract(ext_self)]
+pub trait ExtSelf {
+    fn on_lockup_create(
+        &mut self,
+        lockup_account_id: AccountId,
+        attached_deposit: U128,
+        predecessor_account_id: AccountId,
+    ) -> Promise;
+}
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct LockupFactory {
+    master_account_id: AccountId,
+    lockup_master_account_id: AccountId,
+    whitelist_account_id: AccountId,
+    foundation_account_id: AccountId,
+}
+
+#[derive(Serialize)]
+#[serde(crate = "near_sdk::serde")]
+pub struct LockupArgs {
+    owner_account_id: AccountId,
+    lockup_duration: WrappedDuration,
+    lockup_timestamp: Option<WrappedTimestamp>,
+    transfers_information: TransfersInformation,
+    vesting_schedule: Option<VestingScheduleOrHash>,
+    release_duration: Option<WrappedDuration>,
+    staking_pool_whitelist_account_id: AccountId,
+    foundation_account_id: Option<AccountId>,
+}
+
+
+impl Default for LockupFactory {
+    fn default() -> Self {
+        env::panic(b"LockupFactory should be initialized before usage")
+    }
+}
+
+#[near_bindgen]
+impl LockupFactory {
+    #[init]
+    pub fn new(master_account_id: AccountId,
+               lockup_master_account_id: AccountId,
+               whitelist_account_id: AccountId,
+               foundation_account_id: AccountId) ->
+               Self {
+        assert!(!env::state_exists(), "The contract is already initialized");
+
+        assert!(
+            env::is_valid_account_id(master_account_id.as_bytes()),
+            "The master account ID is invalid"
+        );
+
+        assert!(
+            env::is_valid_account_id(lockup_master_account_id.as_bytes()),
+            "The lockup account ID is invalid"
+        );
+        assert!(
+            env::is_valid_account_id(whitelist_account_id.as_bytes()),
+            "The whitelist account ID is invalid"
+        );
+
+        assert!(
+            env::is_valid_account_id(foundation_account_id.as_bytes()),
+            "The foundation account is invalid"
+        );
+        Self {
+            master_account_id,
+            lockup_master_account_id,
+            whitelist_account_id,
+            foundation_account_id,
+        }
+    }
+
+    /// Returns the foundation account id.
+    pub fn get_foundation_account_id(&self) -> String {
+        self.foundation_account_id.to_string()
+    }
+
+    /// Returns the master account id.
+    pub fn get_master_account_id(&self) -> String {
+        self.master_account_id.to_string()
+    }
+
+    /// Returns the lockup account id.
+    pub fn get_lockup_master_account_id(&self) -> String {
+        self.lockup_master_account_id.to_string()
+    }
```

Comment:
Use `clone` instead of `to_string()` and return `AccountId` to indicate that this is an account 

```suggestion
    pub fn get_foundation_account_id(&self) -> AccountId {
        self.foundation_account_id.clone()
    }

    /// Returns the master account id.
    pub fn get_master_account_id(&self) -> AccountId {
        self.master_account_id.clone()
    }

    /// Returns the lockup account id.
    pub fn get_lockup_master_account_id(&self) -> AccountId {
        self.lockup_master_account_id.clone()
    }
```